Support unsigned int for randint, item, equality, fill, iinfo, tensor

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #116809
* #116810
* #116808
* #116807
* #116806
* __->__ #116805

These are some basic utilities that are often used for testing.

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10

ezyang committed:
Support unsigned int for randint, item, equality, fill, iinfo, tensor

These are some basic utilities that are often used for testing.

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

[ghstack-poisoned]

ezyang committed:
Update on "Support unsigned int for randint, item, equality, fill, iinfo, tensor"


These are some basic utilities that are often used for testing.

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

[ghstack-poisoned]

albanD commented on aten/src/ATen/native/DistributionTemplates.h:
```diff
@@ -2,6 +2,7 @@
 
 #include <ATen/core/Tensor.h>
 #include <ATen/Dispatch.h>
+#include <ATen/Dispatch_v2.h>
```
You can remove the Dispatch.h include above

ezyang replied:
iwyu would be better to include it haha

albanD commented on aten/src/ATen/native/DistributionTemplates.h:
```diff
@@ -110,13 +111,20 @@ static void check_from_to_in_range(int64_t from, int64_t to_inc, caffe2::TypeMet
       WARN_OUT_OF_BOUNDS(to_inc, "to - 1", digits, dtype);
     });
+  } else if (scalar_type == kUInt64) {
+    AT_DISPATCH_V2(scalar_type, "check_random_integral_bounds", AT_WRAP([&]() {
```
What is the benefit of the dispatch macro if you only give it a single dtype? Is it not a no-op?

ezyang replied:
Hmm, actually I guess this is kind of pointless, especially here where I'm not using `scalar_t` lol.

albanD commented on aten/src/ATen/native/DistributionTemplates.h:
```diff
@@ -110,13 +111,20 @@ static void check_from_to_in_range(int64_t from, int64_t to_inc, caffe2::TypeMet
     });
+  } else if (scalar_type == kUInt64) {
+    AT_DISPATCH_V2(scalar_type, "check_random_integral_bounds", AT_WRAP([&]() {
+      // This is annoying: because randint takes int arguments, this means
```
Why can't you use UINT64_T bounds here?
the min will be 0 and the max will be larger than INT64_MAX so it is ok to use.
Or it least to bad implicit casting of "from" and "to_inc" that are problematic?

ezyang replied:
This piece is sooooo annoying.

When you do a comparison between int64_t and uint64_t, the usual arithmetic conversions say that the int64_t value is promoted to unsigned. But this conversion wraps around: if I had -1 as my int64_t, then it will promote to 0xFFFFFFFFFFFFFFFF in uint64_t. This is never the right thing to do.

albanD commented on aten/src/ATen/native/cpu/BinaryOpsKernel.cpp:
```diff
@@ -4,6 +4,7 @@
 #include <cmath>
 
 #include <ATen/Dispatch.h>
+#include <ATen/Dispatch_v2.h>
```
Include above not needed

albanD commented on torch/csrc/TypeInfo.cpp:
```diff
@@ -150,11 +152,19 @@ static PyObject* THPFInfo_min(THPFInfo* self, void*) {
-    return AT_DISPATCH_INTEGRAL_TYPES(self->type, "max", [] {
-      return THPUtils_packInt64(std::numeric_limits<scalar_t>::max());
+    return AT_DISPATCH_IINFO_TYPES(self->type, "max", [] {
+      if (std::is_unsigned<scalar_t>::value) {
```
nit: here and below since we had quite a few lint PRs fixing these
```suggestion
      if (std::is_unsigned_v<scalar_t>) {
```

albanD commented on torch/csrc/utils/python_scalars.h:
```diff
@@ -36,7 +36,8 @@ inline void store_scalar(void* data, at::ScalarType scalarType, PyObject* obj) {
       break;
     case at::kUInt64:
-      *(uint64_t*)data = unpackIntegral<uint64_t>(obj, "uint64");
+      // NB: This doesn't allow implicit conversion of float to int
```
Why the generic unpackIntegral didn't work out here?

ezyang replied:
unpackIntegral unpacks to int64_t which cannot represent the upper half of uint64_t. So the unpackIntegral will fail with an overflow error. You can unpack it directly into uint64_t though, which is what we do below.

ezyang committed:
Update on "Support unsigned int for randint, item, equality, fill, iinfo, tensor"


These are some basic utilities that are often used for testing.

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

[ghstack-poisoned]

ezyang committed:
Update on "Support unsigned int for randint, item, equality, fill, iinfo, tensor"


These are some basic utilities that are often used for testing.

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

[ghstack-poisoned]

ezyang committed:
Update on "Support unsigned int for randint, item, equality, fill, iinfo, tensor"


These are some basic utilities that are often used for testing.

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

[ghstack-poisoned]

albanD reviewed (approved):
Thanks for the update!

ezyang committed:
Update on "Support unsigned int for randint, item, equality, fill, iinfo, tensor"


These are some basic utilities that are often used for testing.

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

[ghstack-poisoned]

ezyang commented:
I ported some more fixes, mostly for CUDA stuff here:

* Turn off reduction upcasting
* Turn off promotion more hard
* Add CUDA equality support
* Add CUDA copy_ support
* Add CUDA index_select support (this one was used by a test, and I figured indexing ops probably are worth it)
* Add CUDA cat support (this doesn't actually net out more kernels because of the opaque dtype trick)

ezyang committed:
Update on "Support unsigned int for randint, item, equality, fill, iinfo, tensor"


These are some basic utilities that are often used for testing.

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

[ghstack-poisoned]
